### PR TITLE
gha: re-enable the release protector

### PR DESCRIPTION
**Pull request by jhchabran**
Follow-up to https://sourcegraph.slack.com/archives/C032Z79NZQC/p1673288833958559?thread_ts=1673284446.955669&cid=C032Z79NZQC 

The idea is that we want to recycle the script we had in place for the 4.0 release that warn others to make sure to acknowledge that their changes are going into the next release. 

We want to avoid scenarios where we're merging changes which would benefit from being tested on S2 and that got instantly released instead, forcing us to issue a patch release if something goes wrong, which is both inconvenient for our customers and the release guild members. 

The check only triggers on the `YYYY-mm-THE_DAY_WE_WANT_TO_FREEZE` date (let's call it freeze date) right now and will: 

- if current date is after the freeze date but before release date. 
  - block PRs without the acknowledging label 
- else 
  - do nothing 

### ➡️ Need to agree on:

- The freeze date
- The release date
- The label 

Once this is done, we can create the label and merge this (after the review ofc).

## Test plan

Extracted the shell part into something configurable: 

(don't pay attention to the garbled emojis, I was pasting the code into a vim session running in a docker container, ran from a terminal within my nvim session).

(if you try this locally, please remember that the BSD date works differently and this need to be ran on a Linux host). 

```sh
# foo.sh
has_label=$1
today=$(date -d "$2" +'%Y%m%d%H%M')

freeze_day=19
release_day=21

current_month=$(date +'%B')
current_year=$(date +'%Y')
freeze_date=$(date -d "${current_month} $freeze_day, ${current_year}" +'%Y%m%d0000')
freeze_date_human=$(date -d "${current_month} $freeze_day, ${current_year}" +'%Y-%m-%d-00:00')
release_date=$(date -d "${current_month} $release_day, ${current_year}" +'%Y%m%d2359')

if [ "$today" -ge "$freeze_date" ] && [ "$today" -lt "$release_date" ]; then
        if [ "$has_label" = "true" ]; then
                echo "�~\~E Label 'i-acknowledge-this-goes-into-the-release' is present"
                exit 0
        else
                echo "�~]~L Label 'i-acknowledge-this-goes-into-the-release' is absent"
                echo "�~_~Q~I We're in the next Sourcegraph release code freeze period. If you are 100% sure your changes should get 
released or provide no risk to the release, add the label your PR with 'i-acknowledge-this-goes-into-the-release'"
                exit 1
        fi
else
        echo "�~_~S~E Not enabled, we're not yet on $freeze_date_human and release code freeze has not started yet." 
        exit 0
fi
```

And ran some tests: (true/false is for the label, second arg is the current date) (freeze day = 19, release_day = 21) 

```
root@aecedf97b54e:~# bash foo.sh false "January 10, 2023"
�📅 Not enabled, we're not yet on 2023-01-19-00:00 and release code freeze has not started yet.
root@aecedf97b54e:~# bash foo.sh false "January 18, 2023"
📅 Not enabled, we're not yet on 2023-01-19-00:00 and release code freeze has not started yet.
root@aecedf97b54e:~# bash foo.sh false "January 19, 2023"
❌ Label 'i-acknowledge-this-goes-into-the-release' is absent
👉 We're in the next Sourcegraph release code freeze period. If you are 100% sure your changes should get released or provide no risk
 to the release, add the label your PR with 'i-acknowledge-this-goes-into-the-release'
root@aecedf97b54e:~# bash foo.sh false "January 20, 2023"
❌ Label 'i-acknowledge-this-goes-into-the-release' is absent
👉 We're in the next Sourcegraph release code freeze period. If you are 100% sure your changes should get released or provide no risk
 to the release, add the label your PR with 'i-acknowledge-this-goes-into-the-release'
root@aecedf97b54e:~# bash foo.sh false "January 21, 2023"
❌ Label 'i-acknowledge-this-goes-into-the-release' is absent
👉 We're in the next Sourcegraph release code freeze period. If you are 100% sure your changes should get released or provide no risk
 to the release, add the label your PR with 'i-acknowledge-this-goes-into-the-release'
root@aecedf97b54e:~# bash foo.sh false "January 22, 2023"
📅 Not enabled, we're not yet on 2023-01-19-00:00 and release code freeze has not started yet.
root@aecedf97b54e:~# bash foo.sh true "January 19, 2023"
✅ Label 'i-acknowledge-this-goes-into-the-release' is present
root@aecedf97b54e:~# bash foo.sh true "January 20, 2023"
✅ Label 'i-acknowledge-this-goes-into-the-release' is present
root@aecedf97b54e:~# bash foo.sh true "January 21, 2023"
✅ Label 'i-acknowledge-this-goes-into-the-release' is present
root@aecedf97b54e:~# bash foo.sh true "January 22, 2023"
📅 Not enabled, we're not yet on 2023-01-19-00:00 and release code freeze has not started yet.
```

### CODE REVIEW

**Comment by BolajiOlajide** [comment]
I see why it's failing, we'll need to do something along t[his line](https://docs.github.com/en/actions/learn-github-actions/finding-and-customizing-actions#adding-an-action-from-the-same-repository).

File: .github/workflows/release-protector.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      ]
+    branches:
+      # only run on branches targeting the `main` branch.
+      - main
+
+jobs:
+  protect-pr:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
```
Comment (BolajiOlajide) [inline .github/workflows/release-protector.yml:22]: I was initially avoiding checking out the repository just to get this working. I've done this before without checking out but just can't remember how I went about it.

This works and I can update to a simpler script that won't need us to checkout the repository later. Checking out takes sometime IIRC

**Review by BolajiOlajide — APPROVED** [review]

File: .github/workflows/release-protector.yml (hunk 1 above)
Comment (jhchabran) [reply .github/workflows/release-protector.yml:22]: Yeah, I thought about this as well, I think we can just backport the script in the action.

Comment (BolajiOlajide) [reply .github/workflows/release-protector.yml:22]: That works . Let's do that instead.
We can move to a go binary much later, something like CODENOTIFY does, so we can reuse across multiple repositories. 
Thanks

**Pull request merged**